rllib/evaluation/collectors/simple_list_collector.py

```diff
@@ -112,49 +128,91 @@ def build(self, view_requirements: Dict[str, ViewRequirement]) -> \
+                requirements dict needed to build an input dict for a ModelV2
+                forward call.
+
         Returns:
             SampleBatch: The built SampleBatch for this agent, ready to go into
                 postprocessing.
         """
 
-        # TODO: measure performance gains when using a UsageTrackingDict
-        #  instead of a SampleBatch for postprocessing (this would eliminate
-        #  copies (for creating this SampleBatch) of many unused columns for
-        #  no reason (not used by postprocessor)).
-
         batch_data = {}
         np_data = {}
         for view_col, view_req in view_requirements.items():
+            # Is an input_dict. Build it using the inference view requirements.
+            if view_req.is_input_dict:
+                batch_data[view_col] = self._get_input_dict(
+                    inference_view_requirements, abs_pos=view_req.abs_pos)
+                continue
+
             # Create the batch of data from the different buffers.
             data_col = view_req.data_col or view_col
+
             # Some columns don't exist yet (get created during postprocessing).
             # -> skip.
             if data_col not in self.buffers:
                 continue
             # OBS are already shifted by -1 (the initial obs starts one ts
             # before all other data columns).
-            shift = view_req.shift - \
-                (1 if data_col == SampleBatch.OBS else 0)
+            obs_shift = -1 if data_col == SampleBatch.OBS else 0
+
+            # Keep an np-array cache so we don't have to regenerate the
+            # np-array for different view_cols using to the same data_col.
+
+            #shift = view_req.data_rel_pos - \
+            #    (1 if data_col == SampleBatch.OBS else 0)
             if data_col not in np_data:
                 np_data[data_col] = to_float_np_array(self.buffers[data_col])
-            # Shift is exactly 0: Send trajectory as is.
-            if shift == 0:
-                data = np_data[data_col][self.shift_before:]
-            # Shift is positive: We still need to 0-pad at the end here.
-            elif shift > 0:
-                data = to_float_np_array(
-                    self.buffers[data_col][self.shift_before + shift:] + [
-                        np.zeros(
-                            shape=view_req.space.shape,
-                            dtype=view_req.space.dtype) for _ in range(shift)
+
+            # Range of indices on time-axis, make sure to create
+            if view_req.data_rel_pos_from is not None:
+                if view_req.batch_repeat_value > 1:
+                    count = int(
+                        math.ceil((len(np_data[data_col]) - self.shift_before)
+                                  / view_req.batch_repeat_value))
+                    repeat_count = (view_req.data_rel_pos_to -
+                                    view_req.data_rel_pos_from + 1)
+                    data = np.asarray([
```

Comment:
Same as above, big confused. Add comments on what these lines of code do

Reply:
done

Reply:
Provided an example.